V2 : adding new class of operations to modeling; modifiers

z3dev opened the pull request:
These changes add a new class of operations to the modeling API, which exposes two functions:
- snap : snap geometries to the precision (epsilon) of the geometry
- generalize : generalize geometries for exporting (including snap, simplify, triangulate, and repair steps)

After working on some changes for Issue #696, it became apparent that some functionality for modifying geometry can be used internally (in preparation for exports), as well as by users. So, there's a new class of operations called 'modifiers'. There are all kinds of functions that simplify, decimate, smooth, etc. geometry. These should be exposed for general use whenever possible.

Why operations called 'modifiers'?
- these are operations as new geometries are produced, just like extrude, union, etc.
- these 'modify' a geometry as a whole using a specific algorithm
- these can be applied to all geometries (in most cases)

### All Submissions:

* [X] Have you followed the guidelines in our Contributing document?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [X] Does your submission pass tests?

z3dev commented:
@SimonClark can you take a quick look at these changes?

this is kind of radical, as the previous versions of JSCAD did not expose these kinds of functions. i think these should be exposed, as there should be no limit on what users can do with the library.

SimonClark commented on packages/modeling/src/operations/modifiers/edges.js:
```diff
@@ -0,0 +1,204 @@
+const poly3 = require('../../geometries/poly3')
+
+// an edge is an array of two vertices
+const addEdge = (edges, edge, polygon) => {
```
Feel like this needs a docblock.  "Add an edge to a polygon, and the polygon to the edge's polygon references. If an equivalent edge is found in the array "edges", then use a reference to that edge instead.".

z3dev replied:
Done

SimonClark commented on packages/modeling/src/operations/modifiers/edges.js:
```diff
@@ -0,0 +1,204 @@
+const addEdge = (edges, edge, polygon) => {
+  let ei = edges.findIndex((element) => {
+    if (element) {
+      if (vec3.equals(element[0], edge[0]) && vec3.equals(element[1], edge[1])) return true
```
These comparisons do not use the JSCAD-wide Epsilon to detect edges that are "close enough".  Should they?

z3dev replied:
Yeah. Edges are expected to be precise, which means 'snap' should be performed beforehand.

SimonClark commented on packages/modeling/src/operations/modifiers/edges.js:
```diff
@@ -0,0 +1,204 @@
+  })
+
+  if (ei < 0) {
+    edge.polygons = [polygon]
```
~~If the edge already has polygons, but is just not found in edges, it will wipe out the edges references.  This seems dangerous if we are exposing this function to unknown uses.~~

Nevermind, this one is not exposed.  Worth annotating as a private method somehow, in general, if a file has multiple exported methods?

SimonClark commented on packages/modeling/src/operations/modifiers/edges.js:
```diff
@@ -0,0 +1,204 @@
+    }
+    return false
+  })
+  if (ei >= 0) {
```
Should we settle on how we say "found in array". Perhaps it should always be explicitly ```ei == -1``` or ```ei != -1```

z3dev replied:
Got it. Fixed.

SimonClark reviewed (commented):
First round of comments, not finished looking but I wanted to publish these...

z3dev commented on packages/modeling/src/maths/vec2/snap.js:
```diff
@@ -0,0 +1,16 @@
+ * @returns {vec2} the resulting vector
+ * @alias module:modeling/maths/vec2.snap
+ */
+const snap = (output, epsilon, input) => {
```
There is...

I'm not convinced that 'optional' parameters in the math classes are a good idea. And these may make JS runtime optimizations impossible.

I have a task for V2.1 to try to prove this. If there's no difference then we can keep the current signitures. If there is a performance improvement then changes are required.

The original [gl-matrix library](https://github.com/toji/gl-matrix) does not have optional parameters, it's always fuction(out, operand, in).

SimonClark replied:
fair, as long as you're on it :)

z3dev commented on packages/modeling/src/maths/vec2/snap.js:
```diff
@@ -0,0 +1,16 @@
+ * @alias module:modeling/maths/vec2.snap
+ */
+const snap = (output, epsilon, input) => {
+  output[0] = Math.round(input[0] / epsilon) * epsilon + 0
```
Basically, calculations can produced either +0 or -0. This little addition makes everything +0.

There's a long long tirad about this on [StackExchange](https://stackoverflow.com/questions/7223359/are-0-and-0-the-same), which discusses all the issues of testing for +/-0 as well as calculations that return different results. It's not a good standard.

SimonClark reviewed (approved):
Kudos!  Lot's of tough math in this one.  I've not been able to take the time to understand double check the math itself, just looked for things that smelled off, and it looks pretty good.

Some minor comments about debugging cruft, and such.  I'll approve, but let you decide to merge or address as you see fit.

z3dev commented on packages/modeling/src/operations/modifiers/edges.js:
```diff
@@ -0,0 +1,204 @@
+}
+
+const removePolygons = (edges, oldedge) => {
+//console.log('removePolygons',oldedge)
```
Done

z3dev commented on packages/modeling/src/operations/modifiers/edges.js:
```diff
@@ -0,0 +1,204 @@
+//console.log('**********OPEN*********')
+//console.log(openedges)
+//console.log('**********OPEN*********')
+console.log('open edges:',openedges.length)
```
Done

z3dev commented on packages/modeling/src/operations/modifiers/edges.js:
```diff
@@ -0,0 +1,204 @@
+  polygons.forEach((polygon) => {
+    addPolygon(edges, polygon)
+  })
+console.log('edges:',edges.length)
```
Done

z3dev commented:
> Kudos! Lot's of tough math in this one. I've not been able to take the time to understand double check the math itself, just looked for things that smelled off, and it looks pretty good.
> 
> Some minor comments about debugging cruft, and such. I'll approve, but let you decide to merge or address as you see fit.

@SimonClark thanks for the review.

the fixing of t-junctions is not straight forward, and requires some tricky algorithms. this isn't perfect yet, and i may try porting the old algorithm as well. SIGH

i cleaned up the code and added comments. it is still kind of weak, so i will add more comments next time through.

SimonClark reviewed (approved):
Okay, I'm happy.
